py/htm/optimization/README.md

```diff
@@ -0,0 +1,150 @@
+In *HTM.core* it is provided in python by `python -m htm.optimization.ae `. 
+
+There are several methods implemented:
+- particle swarm optimization (PSO) by `swarming.py`
+- manual search
+- exhaustive search
+
+## Requirements
+
+- Python 3
+- `ExperimentModule.py` is your python script containing the model to be optimized as
+   well as code to evaluate the models performance.
+- `ExperimentModule.default_parameters = {}`
+   Global dictionary containing all of the parameters to modify.
+   Parameters must be one of the following types: dict, tuple, float, int.
+   Parameters can be nested in multiple levels of dictionaries and tuples.
+   For more details see `htm.optimization.parameter_set`
+- `ExperimentModule.main(parameters=default_parameters, argv=None, verbose=True)`
+   Returns (float) performance of parameters, to be maximized.
+   For example, see file: `py/htm/examples/mnist.py`
+
+## Optimize your model, parameter tuning
+
+Run your experiment with the AE program:
+`$ python3 -m htm.optimization.ae [ae-arguments] ExperimentModule.py [experiment-arguments]`
+
+### Examples
+
+`python -m htm.optimization.ae -n 7 --memory_limit 14 -v --swarming 100  mnist.py`
+
+Uses PSO ("swarming") with 100 particles, 7 CPU threads, upto 14GB RAM, to optimize MNIST results in mnist.py.
+
+`python -m htm.optimization.ae --grid_search "" -n 7 --memory_limit 14 mnist.py` 
+
+Does full exhaustive search for all parameters (the `""`) on `mnist.py` script.
+
+`python -m htm.optimization.ae --grid_search '["potentialPct"]' -n 7 --memory_limit 14 mnist.py`
+
+Tries several different values for the "potentialPct" parameter.
+
+### Manual search
+
+It is possible to manually specify parameters to evaluate.  While the AE program
+is **not** running, append the following to the lab-report file:
+
+```
+================================================================================
+Modification: ['potentialPct'] = 0.66
+Modification: ['columnDimensions'][1] = 123
+```
+
+The 80 equal signs are significant!  The AE program uses them as section dividers.
+
+Each "modification" line is a snippet of python code which is applied to the
+default parameters.
+
+Then run the command: `python -m htm.optimization.ae --parse mnist.py`
+which will read the lab-report file, fill in any missing fields, and write it
+back to the same file, without running any experiments.  After running the AE
+program, the section which we appended to the lab-report will look like:
+
+```
+================================================================================
+Modification: ['columnDimensions'][1] = 123
+Modification: ['potentialPct'] = 0.66
+Hash: FFD67F4E
+Journal: mnist_ae/FFD67F4E.journal
+Attempts: 0
+Notes: 
+```
+
+Notice the "Hash" has been computed for this set of parameters.  The Hash is a
+unique identifier for these parameters.  Use the following command to evaluate
+these parameters:
+
+`python -m htm.optimization.ae --hashes FFD67F4E mnist.py`
```

Comment:
not sure I understand the workflow, but if I do, would be nice if Modifications could be set manually at start, and then the process runs seamlessly (--parse, --hashes)

Reply:
The problem is that you do not know what the hashes are until the lab-report has been parsed.  So first you need to parse the lab-report, then copy-paste the hashes from the lab-report into the command line argument `--hashes`.

In fact, every command to the AE program will parse and write-back the lab-report, it's just convenient to have a command named "parse" which does this and then exits.